src/main/java/io/quarkus/gizmo2/impl/BlockCreatorImpl.java

```diff
@@ -1108,20 +1114,51 @@ public Expr exprToString(final Expr expr) {
             case LONG -> long.class;
             case FLOAT -> float.class;
             case DOUBLE -> double.class;
-            case REFERENCE -> expr.type().isArray() ? switch (TypeKind.from(expr.type().componentType())) {
-                case CHAR -> char[].class;
-                default -> Object.class;
-            } : Object.class;
+            case REFERENCE -> expr.type().isArray() && expr.type().componentType().equals(CD_char) ? char[].class : Object.class;
             default -> throw new IllegalArgumentException("Invalid type for `toString`: " + expr);
         }), expr);
     }
 
+    public Expr arrayHashCode(final Expr expr) {
+        if (!expr.type().isArray()) {
+            throw new IllegalArgumentException("arrayHashCode parameter must be of an array type");
+        }
+
+        ClassDesc componentType = expr.type().componentType();
+        if (componentType.isArray()) {
+            return invokeStatic(MethodDesc.of(Arrays.class, "deepHashCode", int.class, Object[].class), expr);
+        } else {
+            ClassDesc type = TypeKind.from(componentType) == TypeKind.REFERENCE ? CD_Object.arrayType() : expr.type();
+            return invokeStatic(MethodDesc.of(Arrays.class, "hashCode", MethodTypeDesc.of(CD_int, type)), expr);
+        }
+    }
+
     public Expr arrayEquals(final Expr a, final Expr b) {
-        ClassDesc type = switch (TypeKind.from(a.type().componentType())) {
-            case REFERENCE -> CD_Object.arrayType();
-            default -> a.type();
-        };
-        return invokeStatic(MethodDesc.of(Arrays.class, "equals", MethodTypeDesc.of(CD_boolean, type, type)), a, b);
+        if (!a.type().isArray() || !b.type().isArray() || !a.type().equals(b.type())) {
```

Comment:
This isn't strictly true; I could compare a `String[]` and an `Object[]` for equality, right? Maybe it should just check for equality of the `TypeKind` instead.

Also, we have some validation methods in `Item` which are normally invoked from the constructor of `Item` subclasses, but I wonder if they should be centralized (maybe even using JBoss Logging message formatting stuff?).

Reply:
Done. Moved the validation methods from `Item` to a new class `Preconditions` (that's the name I use in SmallRye Fault Tolerance, can use a different one if you prefer). Didn't use the JBoss Logging formatting stuff yet, maybe later?